In [1]:
import pandas as pd
import numpy as np

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
data = pd.read_csv('/content/drive/MyDrive/Rawmart Assessment /Sales.csv')

In [4]:
data.head()

Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
0       1          INDMKB   8/11/2020  11/11/2020             sas          as   
1       2  CA-2020-152156   8/11/2020  11/11/2020    Second Class    CG-12520   
2       3  CA-2020-138688   12/6/2020   16/6/2020    Second Class    DV-13045   
3       4  US-2019-108966  11/10/2019  18/10/2019  Standard Class    SO-20335   
4       5  US-2019-108966  11/10/2019  18/10/2019  Standard Class    SO-20335   

     Customer Name    Segment Country/Region             City  ...  \
0              asa        asa           asas             asas  ...   
1            asasa   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
0     42420.0  INDMKB  FUR-BO-10001798        Furniture    Bookcases   
1     42420.0   South  FUR-CH-10000454        Furniture       Chairs   
2     90036.0    West  OFF-LA-10000240  Office Supplies       Labels   
3     33311.0   South  FUR-TA-10000577        Furniture       Tables   
4     33311.0   South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 21 columns]

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Row ID          9994 non-null   int64  
 1   Order ID        9994 non-null   object 
 2   Order Date      9994 non-null   object 
 3   Ship Date       9994 non-null   object 
 4   Ship Mode       9994 non-null   object 
 5   Customer ID     9994 non-null   object 
 6   Customer Name   9994 non-null   object 
 7   Segment         9994 non-null   object 
 8   Country/Region  9994 non-null   object 
 9   City            9994 non-null   object 
 10  State           9994 non-null   object 
 11  Postal Code     9983 non-null   float64
 12  Region          9994 non-null   object 
 13  Product ID      9994 non-null   object 
 14  Category        9994 non-null   object 
 15  Sub-Category    9994 non-null   object 
 16  Product Name    9994 non-null   object 
 17  Sales           9994 non-null   f

In [6]:
data["Order Date"] = pd.to_datetime(data["Order Date"], dayfirst=True)
data["Ship Date"] = pd.to_datetime(data["Ship Date"], dayfirst=True)


In [7]:
data.isna().sum()

Row ID             0
Order ID           0
Order Date         0
Ship Date          0
Ship Mode          0
Customer ID        0
Customer Name      0
Segment            0
Country/Region     0
City               0
State              0
Postal Code       11
Region             0
Product ID         0
Category           0
Sub-Category       0
Product Name       0
Sales              0
Quantity           0
Discount           0
Profit             0
dtype: int64

In [8]:
data['Postal Code'] = data['Postal Code'].fillna(5401)

In [9]:
data.duplicated().sum()

np.int64(0)

In [10]:
date_errors = data[data['Ship Date'] < data['Order Date']]

In [11]:
date_errors

Empty DataFrame
Columns: [Row ID, Order ID, Order Date, Ship Date, Ship Mode, Customer ID, Customer Name, Segment, Country/Region, City, State, Postal Code, Region, Product ID, Category, Sub-Category, Product Name, Sales, Quantity, Discount, Profit]
Index: []

[0 rows x 21 columns]

In [12]:
data["Ship Mode"].unique()

array(['sas', 'Second Class', 'Standard Class', 'First Class', 'Same Day'],
      dtype=object)

In [13]:
data["Segment"].unique()

array(['asa', 'Consumer', 'Corporate', 'Home Office'], dtype=object)

In [14]:
data["Category"].unique()

array(['Furniture', 'Office Supplies', 'Technology'], dtype=object)

In [15]:
data["Region"].unique()

array(['INDMKB', 'South', 'West', 'Central', 'East'], dtype=object)

In [16]:
data = data[data['Ship Mode'] != 'sas']

In [17]:
data['Ship Mode'].unique()

array(['Second Class', 'Standard Class', 'First Class', 'Same Day'],
      dtype=object)

In [18]:
data.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
1       2  CA-2020-152156 2020-11-08 2020-11-11    Second Class    CG-12520   
2       3  CA-2020-138688 2020-06-12 2020-06-16    Second Class    DV-13045   
3       4  US-2019-108966 2019-10-11 2019-10-18  Standard Class    SO-20335   
4       5  US-2019-108966 2019-10-11 2019-10-18  Standard Class    SO-20335   
5       6  CA-2018-115812 2018-06-09 2018-06-14  Standard Class    BH-11710   

     Customer Name    Segment Country/Region             City  ...  \
1            asasa   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
5  Brosina Hoffman   Consumer  United States      Los Angeles  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
1     42420.0   South  FUR-CH-10000454        Furniture       Chairs   
2     90036.0    West  OFF-LA-10000240  Office Supplies       Labels   
3     33311.0   South  FUR-TA-10000577        Furniture       Tables   
4     33311.0   South  OFF-ST-10000760  Office Supplies      Storage   
5     90032.0    West  FUR-FU-10001487        Furniture  Furnishings   

                                        Product Name     Sales  Quantity  \
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   
5  Eldon Expressions Wood and Plastic Desk Access...   48.8600         7   

   Discount    Profit  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  
5      0.00   14.1694  

[5 rows x 21 columns]

In [19]:
max_date = data['Order Date'].max()
current_year = max_date.year
previous_year = current_year - 1

In [20]:
customers_current = set(
    data.loc[data['Order Date'].dt.year == current_year, 'Customer ID'].dropna().unique()
)
customers_previous = set(
    data.loc[data['Order Date'].dt.year == previous_year, 'Customer ID'].dropna().unique()
)
customers_historic = set(
    data.loc[data['Order Date'].dt.year < current_year, 'Customer ID'].dropna().unique()
)

customer = (
    data.groupby('Customer ID', as_index=False)
        .agg(
            first_order_date=('Order Date', 'min'),
            last_order_date=('Order Date', 'max'),
            order_count=('Order ID', 'nunique'),
            total_spent=('Sales', 'sum')
        )
)

customer['is_active'] = (customer['last_order_date'] >= max_date - pd.Timedelta(days=90)).astype(int)
customer['is_repeat'] = (customer['order_count'] > 1).astype(int)

# New = in current year and not in historic
customer['is_new'] = (
    customer['Customer ID'].isin(customers_current) &
    ~customer['Customer ID'].isin(customers_historic)
).astype(int)

# Churned = in previous year and not in current year
customer['is_churned'] = (
    customer['Customer ID'].isin(customers_previous) &
    ~customer['Customer ID'].isin(customers_current)
).astype(int)


In [21]:
customer.head()

Customer ID first_order_date last_order_date  order_count  total_spent  \
0    AA-10315       2018-03-31      2021-06-29            5     5563.560   
1    AA-10375       2018-04-21      2021-12-11            9     1056.390   
2    AA-10480       2018-05-04      2021-04-15            4     1790.512   
3    AA-10645       2018-06-22      2021-11-05            6     5086.935   
4    AB-10015       2018-02-18      2020-11-10            3      886.156   

   is_active  is_repeat  is_new  is_churned  
0          0          1       0           0  
1          1          1       0           0  
2          0          1       0           0  
3          1          1       0           0  
4          0          1       0           1

In [22]:
active_customers = customer['is_active'].sum()
print(active_customers)

436


In [23]:
new_customers = customer['is_new'].sum()
print(new_customers)

11


In [24]:
total_previous = len(customers_previous)
churn_rate = (customer['is_churned'].sum() / total_previous) * 100
print(churn_rate)

12.539184952978054


In [25]:
retained_count = len(customers_previous.intersection(customers_current))
retention_rate = (retained_count / total_previous) * 100
print(retention_rate)

87.46081504702194


In [26]:
aov = data['Sales'].sum() / data['Order ID'].nunique()
print(aov)

458.56236779796365


In [27]:
ltv = customer['total_spent'].mean()
print(ltv)

2896.5181592686


In [28]:
category_perf = data.groupby('Category').agg({
    'Sales': 'sum',
}).sort_values(by='Sales', ascending=False)

category_perf.head()

Sales
Category                    
Technology       836154.0330
Furniture        741737.8353
Office Supplies  719047.0320

In [29]:
top_sub_cat = data.groupby('Sub-Category')['Sales'].sum().nlargest(10)
top_sub_cat

Sub-Category
Phones         330007.0540
Chairs         328449.1030
Storage        223843.6080
Tables         206965.5320
Binders        203412.7330
Machines       189238.6310
Accessories    167380.3180
Copiers        149528.0300
Bookcases      114618.0363
Appliances     107532.1610
Name: Sales, dtype: float64

In [30]:
top_products = data.groupby('Product Name')['Sales'].sum().nlargest(10)
top_products

Product Name
Canon imageCLASS 2200 Advanced Copier                                          61599.824
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind    27453.384
Cisco TelePresence System EX90 Videoconferencing Unit                          22638.480
HON 5400 Series Task Chairs for Big and Tall                                   21870.576
GBC DocuBind TL300 Electric Binding System                                     19823.479
GBC Ibimaster 500 Manual ProClick Binding System                               19024.500
Hewlett Packard LaserJet 3310 Copier                                           18839.686
HP Designjet T520 Inkjet Large Format Printer - 24" Color                      18374.895
GBC DocuBind P400 Electric Binding System                                      17965.068
High Speed Automatic Electric Letter Opener                                    17030.312
Name: Sales, dtype: float64

In [31]:
customer['recency'] = (max_date - customer['last_order_date']).dt.days

In [32]:
customer = customer.rename(columns={'order_count': 'frequency', 'total_spent': 'monetary'})

In [33]:
customer[['frequency','monetary','recency']].describe().T

count         mean          std    min       25%       50%  \
frequency  793.0     6.316520     2.550885  1.000     5.000     6.000   
monetary   793.0  2896.518159  2628.906521  4.833  1137.616  2256.394   
recency    793.0   146.802018   186.211051  0.000    30.000    75.000   

                75%       max  
frequency     8.000     17.00  
monetary   3785.276  25043.05  
recency     183.000   1165.00

In [34]:
customer['R_Score'] = pd.qcut(customer['recency'], 5, labels=[5,4, 3, 2, 1])

In [35]:
customer['F_Score'] = pd.qcut(customer['frequency'].rank(method='first'), 5, labels=[1,2,3,4,5])

In [36]:
customer['M_Score'] = pd.qcut(customer['monetary'], 5, labels=[1,2,3,4,5])

In [37]:
customer.head()

Customer ID first_order_date last_order_date  frequency  monetary  \
0    AA-10315       2018-03-31      2021-06-29          5  5563.560   
1    AA-10375       2018-04-21      2021-12-11          9  1056.390   
2    AA-10480       2018-05-04      2021-04-15          4  1790.512   
3    AA-10645       2018-06-22      2021-11-05          6  5086.935   
4    AB-10015       2018-02-18      2020-11-10          3   886.156   

   is_active  is_repeat  is_new  is_churned  recency R_Score F_Score M_Score  
0          0          1       0           0      184       2       2       5  
1          1          1       0           0       19       5       5       2  
2          0          1       0           0      259       1       1       3  
3          1          1       0           0       55       3       3       5  
4          0          1       0           1      415       1       1       1

In [38]:
customer['RFM_Segment'] = customer['R_Score'].astype(str) + customer['F_Score'].astype(str)

In [39]:
customer.head()

Customer ID first_order_date last_order_date  frequency  monetary  \
0    AA-10315       2018-03-31      2021-06-29          5  5563.560   
1    AA-10375       2018-04-21      2021-12-11          9  1056.390   
2    AA-10480       2018-05-04      2021-04-15          4  1790.512   
3    AA-10645       2018-06-22      2021-11-05          6  5086.935   
4    AB-10015       2018-02-18      2020-11-10          3   886.156   

   is_active  is_repeat  is_new  is_churned  recency R_Score F_Score M_Score  \
0          0          1       0           0      184       2       2       5   
1          1          1       0           0       19       5       5       2   
2          0          1       0           0      259       1       1       3   
3          1          1       0           0       55       3       3       5   
4          0          1       0           1      415       1       1       1   

  RFM_Segment  
0          22  
1          55  
2          11  
3          33  
4          11

In [40]:
seg_map = {
r'[1-2][1-2]': 'hibernating',
r'[1-2][3-4]': 'at Risk',
r'[1-2]5': 'can\'t_loose',
r'3[1-2]': 'about_to_sleep',
r'33': 'need attention',
r'[3-4][4-5]' :'loyal customers',
r'41': 'promising',
r'51': 'new_customers',
r'[4-5][2-3]': 'potential_loyalists',
r'5[4-5]': 'champions'
}
customer['segment']=customer['RFM_Segment'].replace(seg_map, regex=True)

**Customer Segment Definitions**


    • Champions: Our best and most loyal customers

    • Loyal Customers:Frequent buyers

    • Potential Loyalists: Recent customers who have spent a good amount and bought more than once

    • New Customers: Customers who made their first purchase very recently.

    • Promising: Recent buyers who haven't spent much yet but show potential.

    • Need Attention: Customers with above-average recency and frequency, but they haven't purchased in a while.

    •	About to Sleep: Below-average recency and frequency. We are at risk of losing them.

    • At Risk: These were once frequent, high-spending customers, but they haven't purchased in a long time.

    • Can’t Lose Them: Our former "Champions" who have stopped buying entirely.

    • Hibernating: Our lowest-performing group. They haven't bought in a long time, bought infrequently, and spent very little.




In [41]:
customer.sample(10)

Customer ID first_order_date last_order_date  frequency  monetary  \
483    MA-17995       2018-09-20      2021-06-27          6  2453.284   
452    LC-16930       2018-02-20      2021-11-19          8  4492.946   
15     AD-10180       2018-11-19      2021-12-01          8  6106.880   
31     AH-10210       2018-11-29      2021-12-24          9  4805.344   
281    FC-14335       2018-10-10      2021-06-29          7  3889.366   
86     BG-11695       2018-05-11      2021-11-24          6  1874.172   
765    TS-21430       2018-07-30      2021-07-11          5  2054.140   
461    LH-16900       2018-08-19      2021-12-09          9  2295.332   
615    QJ-19255       2018-03-17      2021-12-05          9  6108.338   
134    CC-12550       2021-06-15      2021-12-03          3   113.832   

     is_active  is_repeat  is_new  is_churned  recency R_Score F_Score  \
483          0          1       0           0      186       2       3   
452          1          1       0           0       41       4       4   
15           1          1       0           0       29       4       4   
31           1          1       0           0        6       5       5   
281          0          1       0           0      184       2       3   
86           1          1       0           0       36       4       3   
765          0          1       0           0      172       2       3   
461          1          1       0           0       21       5       5   
615          1          1       0           0       25       5       5   
134          1          1       1           0       27       4       1   

    M_Score RFM_Segment              segment  
483       3          23              at Risk  
452       5          44      loyal customers  
15        5          44      loyal customers  
31        5          55            champions  
281       4          23              at Risk  
86        3          43  potential_loyalists  
765       3          23              at Risk  
461       3          55            champions  
615       5          55            champions  
134       1          41            promising

In [42]:
customer[['segment','recency','frequency','monetary']].groupby('segment').agg(['mean','count','max']).round()

recency             frequency           monetary        \
                       mean count   max      mean count max     mean count   
segment                                                                      
about_to_sleep         73.0    57   100       4.0    57   5   2487.0    57   
at Risk               223.0   106   719       7.0   106   8   3099.0   106   
can't_loose           211.0    39   410      10.0    39  13   4258.0    39   
champions              13.0    88    26       9.0    88  17   3933.0    88   
hibernating           377.0   171  1165       4.0   171   5   1683.0   171   
loyal customers        56.0   149   103       9.0   149  13   3904.0   149   
need attention         79.0    32   103       6.0    32   7   2914.0    32   
new_customers          14.0    22    26       3.0    22   4   1082.0    22   
potential_loyalists    26.0   112    53       6.0   112   7   2647.0   112   
promising              39.0    17    50       3.0    17   4   1859.0    17   

                              
                         max  
segment                       
about_to_sleep       25043.0  
at Risk              11165.0  
can't_loose          14142.0  
champions            12209.0  
hibernating          19052.0  
loyal customers      14474.0  
need attention       15117.0  
new_customers         3407.0  
potential_loyalists  12873.0  
promising             8954.0

In [44]:
data.to_csv('Rawmart_Data.csv', index=False)

In [45]:
customer.to_csv('Customer_data.csv', index =False)


# Model

In [60]:
features = [
    'recency',
    'frequency',
    'monetary',
    'is_active',
    'is_repeat',
    'is_new'
]

X = customer[features]
y = customer['is_churned']


In [61]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


In [62]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)


LogisticRegression(max_iter=1000)

In [63]:
from sklearn.metrics import confusion_matrix

y_pred = model.predict(X_test)
confusion_matrix(y_test, y_pred)


array([[139,   4],
       [ 12,   4]])

In [64]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.92      0.97      0.95       143
           1       0.50      0.25      0.33        16

    accuracy                           0.90       159
   macro avg       0.71      0.61      0.64       159
weighted avg       0.88      0.90      0.88       159



In [65]:
from sklearn.metrics import roc_auc_score

y_prob = model.predict_proba(X_test)[:,1]
roc_auc_score(y_test, y_prob)


np.float64(0.9685314685314685)

In [66]:

score = model.score(X_test, y_test)

print(f"Accuracy: {score}")

Accuracy: 0.89937106918239


    •	selected RFM-based features and activity flags as they represent customer engagement and purchasing behavior.

    •	Logistic Regression was chosen for its interpretability, stability, and ability to estimate churn probability in a binary classification problem.

    •	The model was evaluated using Recall, F1-score, and ROC-AUC to account for class imbalance and prioritize identifying churned customers

    •	The model achieved an accuracy of ~89.9%, indicating strong overall performance in distinguishing between churned and non-churned customers